resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/localstore/FileSystemResourceManager.java

```diff
@@ -1342,7 +1345,23 @@ public void write(IFile target, byte[] content, IFileInfo fileInfo, int updateFl
 			subMonitor.split(1);
 		}
 		int options = append ? EFS.APPEND : EFS.NONE;
-		store.write(content, options, subMonitor.split(1));
+		try {
+			store.write(content, options, subMonitor.split(1));
+		} catch (CoreException e) {
+			if (append) {
+				throw e;
+			}
+			if (e.getStatus().getCode() != EFS.ERROR_WRITE) {
+				throw e;
+			}
+			IFileStore parent = store.getParent();
+			IFileInfo parentInfo = parent.fetchInfo();
+			if (!parentInfo.exists()) {
+				parent.mkdir(EFS.NONE, null);
+				// try again
+				store.write(content, options, null);
+			}
```

Comment:
Would it be possible to handle this within the store implementation?
This would save throwing and catching the CoreException. But this would probably require an extra options-flag.

According to this article creating exceptions is still surprisingly expensive:
https://www.baeldung.com/java-exceptions-performance

When I compare it to the numbers I found for https://github.com/eclipse-platform/eclipse.platform/pull/1422#issuecomment-2184161944 I see a difference of three magnitudes.

I have to admit that I would not have expected that throwing an exception is almost a thousand times slower than fetching file infos from the drive. So I'm not sure if I miss-interpret something and of curse comparing benchmarks from different machines is usually not meaningful.
But I think it would be good to have some form of benchmark if the new solution is net better than the old approach in an average workspace.
I assume that if I for example run a CLEAN+FULL_BUILD that all folders in the classpath output folder are deleted and re-created when the first class file was compiled? Overall these are many folders that are then missing on first write.

Reply:
If throwing/catching exceptions is indeed that expensive, can't we just reuse the fetched file-info instead?

Reply:
> According to this article creating exceptions is still surprisingly expensive:

I also mentioned that last time with the "NIO vs native" discussion, the NIO is*only* noticeable slower when ist came to non exiting files because there Exceptions are used and if you have a large number of exception it makes a difference, but in general one should not use exceptions to control the flow anyways

Reply:
> According to this article creating exceptions is still surprisingly expensive:
> https://www.baeldung.com/java-exceptions-performance

Which numbers do you read? Be careful: that Article measures a loop of LIMIT invocations (which is bad practice) and gives numbers of that. It also does not mention that the time actually highly depends on the depth on the stacktrace elements, So it just can't give absolute numbers.

> I assume that if I for example run a CLEAN+FULL_BUILD that all folders in the classpath output folder are deleted and re-created when the first class file was compiled? Overall these are many folders that are then missing on first write.

Put a breakpoint in the catch clause and run the action you mention => it never stops. Because it would require the absurd conditions given in the added testcase: The folder must exit in workspace but not exist in FileSystem. That only happens if the folder is deleted without using the resource API.
So this is a non-issue for good plugins using the resource API only.

Reply:
> Would it be possible to handle this within the store implementation?

Only by changing that API which is not the plan for this PR and would have use only for the unlikely use case.

Reply:
> > According to this article creating exceptions is still surprisingly expensive:
> > https://www.baeldung.com/java-exceptions-performance
> 
> Which numbers do you read? Be careful: that Article measures a loop of LIMIT invocations (which is bad practice) and gives numbers of that. It also does not mention that the time actually highly depends on the depth on the stacktrace elements, So it just can't give absolute numbers.

I read the number of the `ExceptionBenchmark.throwAndCatchException` benchmark.
But you are right, this is done 10_000 times in a loop. No clue why this is done since JMH can measure in µs.

> 
> > I assume that if I for example run a CLEAN+FULL_BUILD that all folders in the classpath output folder are deleted and re-created when the first class file was compiled? Overall these are many folders that are then missing on first write.
> 
> Put a breakpoint in the catch clause and run the action you mention => it never stops. Because it would require the absurd conditions given in the added testcase: The folder must exit in workspace but not exist in FileSystem. That only happens if the folder is deleted without using the resource API. So this is a non-issue for good plugins using the resource API only.

OK, I verified that. But please be more descriptive in your PR and commit message what you mean with 'normal case'. Because it is not clear that it can only happen if the the workspace resource store out of sync with the file-system and since the parent existence check for the resources system happens somewhere else also not obvious from the changed code.

Then I'm fine with this.